kyo-prelude/shared/src/main/scala/kyo2/Env.scala

```diff
@@ -0,0 +1,62 @@
+package kyo2
+
+import java.awt.Window.Type
+import kyo.Tag
+import kyo2.*
+import kyo2.kernel.*
+import scala.util.NotGiven
+
+sealed trait Env[+R] extends RuntimeEffect[TypeMap[R]]
+
+object Env:
+
+    private inline def erasedTag[R] = Tag[Env[Any]].asInstanceOf[Tag[Env[R]]]
+
+    inline def get[R](using inline tag: Tag[R]): R < Env[R] =
+        use[R](v => v)
+
+    def run[R >: Nothing: Tag, T, S, RS, RR](env: R)(value: T < (Env[RS] & S))(
+        using
+        HasEnv[R, RS] { type Remainder = RR },
+        Frame
+    ): T < (S & RR) =
+        runTypeMap(TypeMap(env))(value)
+
+    def runTypeMap[R >: Nothing, T, S, RS, RR](env: TypeMap[R])(value: T < (Env[RS] & S))(
+        using
+        HasEnv[R, RS] { type Remainder = RR },
+        Frame
+    ): T < (S & RR) =
+        RuntimeEffect.handle(erasedTag[R], env, _.union(env))(value).asInstanceOf[T < (S & RR)]
+
+    class UseOps[R >: Nothing](ign: Unit) extends AnyVal:
+        inline def apply[A, S](inline f: R => A < S)(
+            using tag: Tag[R]
+        ): A < (Env[R] & S) =
+            def loop(): A < (Env[R] & S) =
+                RuntimeEffect.suspend(erasedTag[R]) { map =>
+                    val value = map.getOrElse(null)(using tag.erased).asInstanceOf[R]
+                    if isNull(value) then loop()
+                    else f(value)
+                }
+            end loop
+            loop()
```

Comment:
Why is the loop now necessary?

Reply:
Sorry, it's a leftover from debugging. Fixed.